### DISCUSSION TITLE
Stray /v1/account/999/_unix/_token in logs

### DISCUSSION THREAD

**Discussion by dvv** [kanidm/kanidm#2522] [Q&A] [answered]
Hi!

I wonder how to get rid of the subj?

```
/v1/account/999/_unix/_token
...
/v1/account/996/_unix/_token
```

```sh
grep 998 /etc/group /etc/passwd
```
is empty.

Where to look at?

TIA

**Comment by yaleman** [comment]
Are you trying to figure out what's causing those logs? It'd be handy if you explained a little more 😄  Those IDs could be ephemeral ones, most likely, it'll be a kanidm-unixd instance trying to query the token endpoint either way... try clearing the auth cache and see if it continues to query.

**Comment by Firstyear** [comment]
The unix resolver will check /etc/group and /etc/passwd first - if either of them own that uid/gid number, then kanidm unixd won't resolve and skips.

If kanidm unixd has not cached the result, it will request it from the the server which responds with a user/group account, or a not-found. The unixd resolver *can* cache not-founds.

The mostly likely scenario (and what I see *a lot*) is spammers/ssh bruteforce, attempting to login to accounts that don't exist. That's where we see the majority of unknown lookups from. There is nothing "harmful" about them, Kanidm won't let them login or proceed, but they can be annoying as you note. 

On the system with kanidm-unixd, check your /etc/nsswitch.conf and see what's there, but otherwise as @yaleman said it could be order of modules causing sysusers.d to be after kani, or something else.

**Comment by dvv** [comment]
I see. The `/etc/nsswitch.conf` is modified just as per the book and I believe the order of resolvers is right:
```
passwd:         files systemd kanidm
group:          files systemd kanidm
```

The portion of kanidm-unixd logs (expired cache):
```
DEBUG    🐛 [debug]: Accepted connection
DEBUG    🐛 [debug]: nssaccountbyuid req
DEBUG    🐛 [debug]: get_usertoken
DEBUG    🐛 [debug]: online expired, refresh cache
DEBUG    perform_get_request [ 82.1µs | 100.00% ] dest: "/v1/account/996/_unix/_token"
DEBUG    ┕━ 🐛 [debug]: opid -> "f9b9861f-690c-4598-8c73-7caffc291035"
DEBUG    🐛 [debug]: entry has been removed or is no longer a valid posix account | opid: "f9b9861f-690c-4598-8c73-7caffc291035"
DEBUG    🐛 [debug]: token -> None
DEBUG    🐛 [debug]: flushed response!
DEBUG    🐛 [debug]: Disconnecting client ...
```
Valid cache:
```
DEBUG    🐛 [debug]: Accepted connection
DEBUG    🐛 [debug]: nssaccountbyuid req
DEBUG    🐛 [debug]: get_usertoken
DEBUG    🐛 [debug]: online valid, returning cached item
DEBUG    🐛 [debug]: token -> None
DEBUG    🐛 [debug]: flushed response!
DEBUG    🐛 [debug]: Disconnecting client ...
```

Do we have means to control `not-found` cache duration? I would like to make it hours, being able to explicitly restart kanidm-unixd to reset the cache manually.
Can we blacklist some (range of?) of uids/gids completely at `/etc/kanidm/unixd` level?
May be kanidm-unixd resolver should/could silently ignore request for names not ending with kanidm domain, as an option?

Another option.
Kanidm logs:
```
INFO     handle_request [ 456µs | 0.00% / 100.00% ]
INFO     ┕━ request [ 456µs | 20.63% / 100.00% ] method: GET | uri: /v1/account/996/_unix/_token | version: HTTP/1.1
INFO        ┝━ account_id_unix_token [ 362µs | 1.66% / 79.37% ]
INFO        │  ┕━ handle_internalunixusertokenread [ 354µs | 33.55% / 77.71% ]
INFO        │     ┝━ validate_client_auth_info_to_ident [ 201µs | 44.16% ]
INFO        │     ┕━ ｉ [info]: Error resolving 998 as gidnumber continuing ... | event_tag_id: 3 | err: NoMatchingEntries
WARN        ┕━ 🚧 [warn]:  | latency: 471.067µs | status_code: 404 | kopid: "7bb6ca23-1b04-45c8-8f63-b6c07f35bd99" | msg: "client error"
```
Here we treat `status_code: 404` as "client error" with elevated log level "WARN".
If such misses were considered as "INFO" (that is perfectly valid and expectable not to found an exact needle in the unknown heap) I could just raise log_level in `/etc/kanidm/server.toml`.

What do you think?

**Reply by Firstyear** [reply, marked as answer]
Yes, your resolver order looks to be correct.

So that means something on the system is trying to lookup id 996. Is there anything in the ssh/auth logs about it? 

Either way we can probably make the 404 an info not warn then.

**Reply by dvv** [reply]
I believe we can not separate such mistaken/malicious requests so the best is to not treat them as warns. Thank you.

**Reply by Firstyear** [reply]
https://github.com/kanidm/kanidm/issues/2526

**Reply by dvv** [reply]
Now I know these requests come from gitlab container. I wonder why it at all is querying the host's nss?

**Reply by dvv** [reply]
It's from container's `su - git` et al.

**Comment by dvv** [comment]
The strangest record I saw so far is for `/v1/account/root/_unix/_token` being `root` in `/etc/passwd` for sure.

**UPD**: I now believe it's `sudo` looking up for the account info, so this particular record is fine.

**Comment by dvv** [comment]
It turned out it's `su` in sibling docker container causes this. I wonder if it's normal and how to get rid of that. I can understand container queries for host' hosts but not passwd/group )
